aggregator/list.go

```diff
@@ -0,0 +1,343 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package aggregator
+
+import (
+	"bytes"
+	"container/list"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3metrics/metric"
+	"github.com/m3db/m3metrics/metric/aggregated"
+	"github.com/m3db/m3metrics/policy"
+	"github.com/m3db/m3metrics/protocol/msgpack"
+	"github.com/m3db/m3x/clock"
+	"github.com/m3db/m3x/log"
+)
+
+var (
+	errListClosed  = errors.New("metric list is closed")
+	errListsClosed = errors.New("metric lists are closed")
+)
+
+type encodeFn func(mp aggregated.ChunkedMetricWithPolicy) error
+
+// MetricList stores aggregated metrics at a given resolution
+// and flushes aggregations periodically
+// TODO(xichen): add metrics
+type MetricList struct {
+	sync.RWMutex
+
+	opts         Options
+	nowFn        clock.NowFn
+	log          xlog.Logger
+	timeLock     *sync.RWMutex
```

Comment:
Yeah I cache it here so that we don't have to call `opts.TimeLock()` on every tick, but I think either way is probably okay because the number of metric lists is small, and we don't tick that often (on the order of seconds)